File: python/binresponses/montebinresp.c
#include "presto.py"

/* #define USEMPI */

#ifdef USEMPI
#include "mpi.h"
#endif

void estimate_rz(psrparms psr, double T, double eo, double *r, double *z){
  double *e, *zs, *rs, dt = 1.0, avgr=0.0, avgz=0.0;
  int ii, numpts;

  numpts = (int)(T + 1);
  e = dorbint(eo, numpts, dt, psr->orb);
  rs = gen_dvect(numpts);
  zs = gen_dvect(numpts);
  for (ii=0 ; ii<numpts ; ii++){
    rs[ii] = e[ii];
    zs[ii] = e[ii];
  }
  E_to_z(zs, numpts, psr.p, T, psr->orb);
  E_to_p(rs, numpts, psr.p, psr->orb);
  for (ii=0 ; ii<numpts ; ii++){
    rs[ii] = T * ( 1.0 / psr.p - 1.0 / rs[ii]);
  }
  for (ii=0 ; ii<numpts ; ii++){
    avgr += rs[ii];
    avgz += zs[ii];
  }
  *r = avgr / numpts;
  *z = avgz / numpts;
  free(e);
  free(rs);
  free(zs);
}

double mass_funct2(double mp, double mc, double i){
  return pow(mc * sin(i), 3.0) / pow(mc + mp, 2.0);
}

double asini_c(double pb, double mf){
  return pow((mf * pb * pb / 8015123.37129), (1.0 / 3.0));
}

int main(int argc, char *argv[]){

  /* Admin parameters */
  char outdir[50] = "/home/ransom";
  char outnm[50] = "montebinresp";
  double pmass = 1.35;
  double cmass = 0.3;
  double ecc = 0.0;
  int orbsperpt = 10;

  /* Simulation parameters */
  int numTbyPb = 50;
  double minTbyPb = 0.01;
  double maxTbyPb = 10.0;
  int numppsr = 50;
  double minppsr = 0.0005;
  double maxppsr = 5.0;
  char ctype[3] = "WD";
  double Pb = 7200.0;
  double dt = 0.0001;
  char searchtype[10] = "ffdot";
  double maxTbyPb_ffdot = 1.0;
  double minTbyPb_sideband = 0.3;
  double fftlen_shortffts = 0.05;

  /* Normal variables */
  int ii, jj, kk, ll, itmp, myid=0, numprocs=1, psr_numbins, datalen, width;
  double dtmp, T, xb, eb, wb, tp, pwr, avgpwr, orbf, orbi, eo, tryr, tryz;
  char outfilenm[200];
  fcomplex *psr_resp;
  psrparams psr;
  FILE *outfile;

#ifdef USEMPI
  MPI_Init(&argc,&argv);
  MPI_Comm_size(MPI_COMM_WORLD,&numprocs);
  MPI_Comm_rank(MPI_COMM_WORLD,&myid);
#endif

  sprintf(outfilenm, "%s/%s_%d_%s_%s.out", \
	  outdir, outnm, myid, searchtype, ctype);
  printf("Output file is '%s'.\n", outfilenm);
  outfile = chkfopen(outfilenm, "w");

  TbyPb = gen_dvect(numTbyPb);
  logTbyPb = gen_dvect(numTbyPb);
  ppsr = gen_dvect(numppsr);
  logppsr = gen_dvect(numppsr);

  for (ii=0 ; ii<numTbyPb ; ii++){
    dtmp = (log(maxTbyPb) - log(minTbyPb)) / (numTbyPb - 1);
    logTbyPb[ii] = log(minTbyPb) + dtmp * ii;
    TbyPb[ii] = exp(logTbyPb[ii]);
  }
  for (ii=0 ; ii<numppsr ; ii++){
    dtmp = (log(maxppsr) - log(minppsr)) / (numppsr - 1);
    logppsr[ii] = log(minppsr) + dtmp * ii;
    ppsr[ii] = exp(logppsr[ii]);
  }

  /* The Simulation loops */

  /* Loop over T / Porb */
  for (ii=0 ; ii<numTbyPb ; ii++){
    T = Pb * TbyPb[ii];
    xb = asini_c(Pb, mass_funct2(pmass, cmass, pi / 3.0));

    /* Loop over ppsr */
    for (jj=0 ; jj<numppsr ; jj++){
      if !(jj % numprocs == myid) continue;
      else {
	pwr = 0.0;
	avgpwr = 0.0;
	if (((strcmp(searchtype, 'ffdot') == 0) &&
	     TbyPb[ii] < maxTbyPb_ffdot) ||
	    ((strcmp(searchtype, 'sideband') == 0) &&
	     TbyPb[ii] > minTbyPb_sideband) ||
	    (strcmp(searchtype, 'shortffts') == 0)){

	  /* Loop over the number of tries per point */
	  for (kk=0 ; kk<orbsperpt ; kk++){
	    if (ecc == 0.0){
	      wb = 0.0;
	      tp = kk * Pb / orbsperpt;
	    } else {
	      orbf = modf(kk / sqrt(orbsperpt), &orbi);
	      orbi /= sqrt(orbsperpt);
	      wb = orbf * 180.0;
	      tp = Pb * orbi;
	    }
	    psr.p = ppsr[jj];
	    psr.orb.p = Pb;
	    psr.orb.x = xb;
	    psr.orb.e = ecc;
	    psr.orb.t = wb;
	    psr.orb.w = tp;
	    psr_numbins = 2 * bin_resp_halfwidth(psr.p, T, psr.orb);
	    psr_resp = gen_bin_response(0.0, 1, psr.p, T, psr.orb,
					psr_numbins);

	    if (strcmp(searchtype,'ffdot')==0){
	      datalen = next2_to_n(psr_numbins * 2)
		if (datalen < 512) datalen = 512;
	      data = gen_cvect(datalen);
	      itmp = (len(data) - len(psr_resp)) / 2
	      for (ll=0 ; ll<psr_numbins ; ll++)
		data[ll+itmp] = psr_resp[ll];
	      eo = keplers_eqn(psr.orb.t, psr.orb.p, psr.orb.e, 1.0e-14)
	      estimate_rz(psr, T, eo, &tryr, &tryz);
	      tryr = tryr + datalen / 2.0;
	      width = 201;
	      ffd = ffdot_plane(data, tryr, 0.5, width, tryz, 2.0, width)
                        maxarg = argmax(spectralpower(ffd.flat))
                        peakr = ((maxarg % width) * 0.5 +
                                 int(tryr - (width * 0.5) / 2.0))
                        peakz = ((maxarg / width) * 2 +
                                 tryz - (width * 2) / 2)
                        if showplots:
                            show_ffdot_plane(data, tryr, tryz)
                            Pgplot.nextplotpage()
                        if debugout:
                            print ' peakr = '+`peakr`+' peakz = '+`peakz`
                        [pows[kk], rmax, zmax, rd] = \
                                   maximize_rz(data, peakr, peakz, \
                                               norm=1.0)
                        if debugout:
                            print ' rmax = '+`rmax`+' zmax = '+`zmax`
                    elif searchtype == 'sideband':
                        psr_pows = spectralpower(psr_resp)
                        data = zeros(next2_to_n(psr_numbins * 2), 'd')
                        data[0:psr_numbins] = psr_pows
                    if debugout:
                        print `x`+'  '+`y`+'  '+`TbyPb[x]`+'  ',
                        print `ppsr[y]`+'  '+`pows[kk]`
                    tim = clock() - stim
                    if debugout:
                        print 'Time for this point was ',tim, ' s.'
            if debugout:
                print `x`+'  '+`y`+'  '+`TbyPb[x]`+'  ',
                print `ppsr[y]`+'  '+`average(pows)`+'  ',
                print `max(pows)`+'  '+`min(pows)`
            file.write(`x`+'  '+`y`+'  '+`TbyPb[x]`+'  ')
            file.write(`ppsr[y]`+'  '+`average(pows)`+'  ')
            file.write(`max(pows)`+'  '+`min(pows)`+'\n')
            file.flush()
  free(TbyPb);
  free(logTbyPb);
  free(ppsr);
  free(logppsr);
#ifdef USEMPI
  MPI_Finalize();
#endif
      }

File: python/binresponses/montebinresp.py
import string, random, sys, cPickle
from math import *
from Numeric import *
from miscutils import *
from presto import *
from orbitstuff import *

# Some admin variables
parallel = 0          # True or false
showplots = 1         # True or false
debugout = 1          # True or false
outfilenm = '/home/ransom/montebinresp_saves'
orbsperpt = 100       # Number of orbits to test
pmass = 1.35          # Pulsar mass in solar masses
cmass = {'WD': 0.3, 'NS': 1.35, 'BH': 10.0}  # Companion masses to use
ecc = {'WD': 0.0, 'NS': 0.6, 'BH': 0.6}      # Eccentricities to use

# Simulation parameters
numTbyPb = 100        # The number of points along the x axis
minTbyPb = 0.01       # Minimum Obs Time / Orbital Period
maxTbyPb = 10.0       # Maximum Obs Time / Orbital Period
numppsr = 100         # The number of points along the y axis
minppsr = 0.0005      # Minimum pulsar period (s)
maxppsr = 5.0         # Maximum pulsar period (s)
ctype = 'WD'          # The type of binary companion: 'WD', 'NS', or 'BH'
dt = 0.0001           # The duration of each data sample (s)
searchtype = 'ffdot'  # One of 'ffdot', 'sideband', 'shortffts'
maxTbyPb_ffdot = 1.0
minTbyPb_sideband = 0.3
fftlen_shortffts = 0.05

##################################################
# You shouldn't need to edit anyting below here. #
##################################################

def psrparams_from_list(pplist):
    psr = psrparams()
    psr.p = pplist[0]
    psr.orb.p = pplist[1]
    psr.orb.x = pplist[2]
    psr.orb.e = pplist[3]
    psr.orb.w = pplist[4]
    psr.orb.t = pplist[5]
    return psr

def estimate_rz(psr, T, eo=0.0):
    """
    estimate_rz(psr, T, eo=0.0):
        Return an estimate of the average Fourier frequency ('r')
        and Fourier F-dot ('z') values in bins given the psrparams
        'psr', the length of the observation 'T' in sec,  and the
        inital Eccentric anomaly 'eo'.
    """
    import LeastSquares
    dt = 1.0
    e = make_orbit(psr, int(T+1.0), dt, eo)
    z = z_from_e(e, psr, T)
    r = T / p_from_e(e, psr)
    return (average(r), average(z))
    
outfilenm = outfilenm+'_'+searchtype+'.txt'

# Calculate the values of our X and Y axis points
logTbyPb = span(log(minTbyPb), log(maxTbyPb), numTbyPb)
logppsr = span(log(minppsr), log(maxppsr), numppsr)
TbyPb = exp(logTbyPb)
ppsr = exp(logppsr)
# Adjust the number of points in the time series so as to
# keep the ppsr in the center of the FFT freq range
# (i.e. fpsr = 1 / ppsr = Nyquist freq / 2)
N = 4 / ppsr
T = N * dt

# Figure out our environment 
if showplots:
    import Pgplot
if parallel:
    import mpi
    from mpihelp import *
    myid = mpi.comm_rank()
    numprocs = mpi.comm_size()
else:
    myid = 0
    numprocs = 1

# Open a file to save each orbit calculation
file = open(outfilenm,'w')

# The Simulation loops

# Loop over T / Porb
for x in range(numTbyPb):
    Pb = T[x] / TbyPb[x]
    xb = asini_c(Pb, mass_funct2(pmass, cmass[ctype], pi / 3.0))
    eb = ecc[ctype]
    # Loop over ppsr
    for y in range(numppsr):
        # Each processor calculates its own point
        if not (y % numprocs == myid):  continue
        else:
            avgpow = 0.0
            if searchtype == 'ffdot':
                if Pb > maxTbyPb_ffdot:
                    print `x`+'  '+`y`+'  '+`TbyPb[x]`+'  ',
                    print `ppsr[y]`+'  '+`avgpow`
                    file.write(`x`+'  '+`y`+'  '+`TbyPb[x]`+'  '+
                               `ppsr[y]`+'  '+`avgpow`)
                    continue
            elif searchtype == 'sideband':
                if Pb < minTbyPb_sideband:
                    print `x`+'  '+`y`+'  '+`TbyPb[x]`+'  ',
                    print `ppsr[y]`+'  '+`avgpow`
                    file.write(`x`+'  '+`y`+'  '+`TbyPb[x]`+'  '+
                               `ppsr[y]`+'  '+`avgpow`)
                    continue
            # Loop over the number of tries per point
            for ct in range(orbsperpt):
                if (eb == 0.0):
                    wb, tp = 0.0, 0.0
                else:
                    (orbf, orbi)  = modf(ct / sqrt(orbsperpt))
                    orbi = orbi / sqrt(orbsperpt)
                    wb, tp = orbf * 180.0, Pb * orbi
                psr = psrparams_from_list([ppsr[y], Pb, xb, eb, wb, tp])
                print numTbyPb, psr.p, psr.orb.p, psr.orb.x
                psr_halfwidth = bin_resp_halfwidth(psr.p, psr.orb)
                psr_numbins = 2 * numbetween * psr_halfwidth
                psr_resp = gen_bin_response(0.0, 1, psr.p, T[x], psr.orb, \
                                            psr_numbins)
                if searchtype == 'ffdot':
                    # The following places the nominative psr freq
                    # in bin # ~len(searchdata) / 2
                    search = zeros(next2_to_n(psr_numbins * 2), 'F')
                    lo = (len(search) - len(psr_resp)) / 2
                    hi = lo + len(psr_resp)
                    search[lo:hi] = psr_resp
                    eo = keplars_eqn(psr.orb.t, psr.orb.p, psr.orb.e, 1.0e-14)
                    (tryr, tryz) = estimate_fdot(psr, T[x], eo)
                    tryr = T[x] / psr.p - tryr + len(searchdata) / 2.0
                    nrz = 201
                    ffd = ffdot_plane(search, tryr, 0.5, nrz, tryz, 2.0, nrz)
                    maxarg = argmax(spectralpower(ffd.flat))
                    [maxpow, rmax, zmax, rd] = maximize_rz(search,
                                                           maxarg % nrz,
                                                           maxarg / nrz)
                    avgpow = avgpow + maxpow
                    if debugout:
                        print `x`+'  '+`y`+'  '+`TbyPb[x]`+'  ',
                        print `ppsr[y]`+'  '+`maxpow`
                        file.write(`x`+'  '+`y`+'  '+`TbyPb[x]`+'  '+
                                   `ppsr[y]`+'  '+`maxpow`)                    
                elif searchtype == 'sideband':
                    psr_pows = spectralpower(psr_resp)
                    search = zeros(next2_to_n(psr_numbins * 2), 'd')
                    search[0:psr_numbins] = psr_pows
            avgpow = avgpow / orbsperpt
            print `x`+'  '+`y`+'  '+`TbyPb[x]`+'  ',
            print `ppsr[y]`+'  '+`avgpow`
            file.write(`x`+'  '+`y`+'  '+`TbyPb[x]`+'  '+
                       `ppsr[y]`+'  '+`avgpow`)                    
            # Save our most recent orbit and search information
#            cPickle.dump(widths[i], file, 1)
file.close()
